src/usb.c

```diff
@@ -0,0 +1,1619 @@
+    }
+
+  if (sender_info->sender_state != USB_SENDER_STATE_ACQUIRING_DEVICES || sender_info->acquiring_devices == NULL)
+    {
+      /* If the request was cancelled in some way. This would happen by calling
+       * FinishAcquireDevices after the user denied the permission.
+       */
+      sender_info->sender_state = USB_SENDER_STATE_DEFAULT;
+      g_clear_pointer (&sender_info->acquiring_devices, g_ptr_array_unref);
+
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                             "There is no device being acquired");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  /* We should never trigger these asserts. */
+  g_assert (sender_info->sender_state == USB_SENDER_STATE_ACQUIRING_DEVICES);
+  g_assert (sender_info->acquiring_devices != NULL);
+
+  fds = g_unix_fd_list_new ();
+
+  g_variant_builder_init (&results_builder, G_VARIANT_TYPE ("a(sa{sv})"));
+
+  accessed_devices = 0;
+  while (accessed_devices < MAX_DEVICES &&
+         sender_info->acquiring_devices->len > 0)
+    {
+      g_autoptr(UsbDeviceAcquireData) access_data = NULL;
+      g_autoptr(GError) error = NULL;
+      xdp_autofd int fd = -1;
+      GVariantDict dict;
+      GUdevDevice *device;
+      UsbOwnedDevice *owned_device = NULL;
+      const char *device_file;
+      int fd_index;
+
+      g_variant_dict_init (&dict, NULL);
+
+      access_data = g_ptr_array_steal_index (sender_info->acquiring_devices, 0);
+      device = g_hash_table_lookup (self->ids_to_devices, access_data->device_id);
+
+      if (!device)
+        {
+          g_variant_dict_insert (&dict, "success", "b", FALSE);
+          g_variant_dict_insert (&dict, "error", "s", _("Device not available"));
+          g_variant_builder_add (&results_builder, "(s@a{sv})",
+                                 access_data->device_id,
+                                 g_variant_dict_end (&dict));
+          continue;
+        }
+
+      device_file = g_udev_device_get_device_file (device);
+      g_assert (device_file != NULL);
+
+      /* Can the app even request this device? */
+      if (!usb_sender_info_match_device (sender_info, device))
+        {
+          g_variant_dict_insert (&dict, "success", "b", FALSE);
+          g_variant_dict_insert (&dict, "error", "s", _("Not allowed"));
+          g_variant_builder_add (&results_builder, "(s@a{sv})",
+                                 access_data->device_id,
+                                 g_variant_dict_end (&dict));
+          continue;
+        }
+
+      /* Check we haven't already acquired the device */
+      owned_device = g_hash_table_lookup (sender_info->owned_devices, access_data->device_id);
+      if (!owned_device)
+        {
+          fd = open (device_file, access_data->writable ? O_RDWR : O_RDONLY);
```

Comment:
Why would opening the file fail? Is udev already telling us that we should be able to open the device node (i.e. we have permission to do it) or is do we discover about bad permissions just here?

Reply:
many reason. It's IO nothing is guaranteed. The device could have disappear, could be unitialized. If it can go wrong it will go wrong.

Reply:
Right, but that means we'll show USB devices to the portal clients, letting users go through the UI to approve access, just to then eventually notice that we actually can't make the device available. Would be much nicer if we didn't advertise those devices in the first place.

Reply:
there is not a lot we can do. We can call access on the device `is_gudev_device_suitable()` and return `FALSE` if the call fail. That will check for existance and read permission. At that leve we don't know if we want the write permissions.

Reply:
And we already check in `handle_acquire_devices()` as well.

Still a lot of race condition can occur it acquiring the device.

Reply:
Yes, obviously there can always be cases where opening will fail because this is inherently racy, but why would we want to advertise devices which we cannot open because we have no read permission? The code right now figures out that it won't be able to open it, advertises it anyway and then fails when a client tries to acquire it.

Reply:
we no longer do. the change just hadn't been pushed yet.

Reply:
Maybe I'm blind but I can't seem to find it? Are you sure you pushed the latest version now?

Reply:
https://github.com/flatpak/xdg-desktop-portal/pull/1354/files#diff-4ce7d34e66e5ca9b52f8e5666f0d4efe11e48278f34a7cdba6df390c14012a27R163

Which mean if it can't access it it won't list it. So open won't be attempted. In case of race condition, ie the device is no longer accessible between the list and this, then this will error normally.

Reply:
Ah, I missed that one. Perfect!

Reply:
actually it caused a serious regression as it didn't detect removals, and a few other things.